File: wltbase/eip712.go
package wltbase

import (
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"math/big"
	"sort"
	"strconv"
	"strings"

	"golang.org/x/crypto/sha3"
)

var (
	// maxUint256 is type(uint256).max — the canonical "unlimited"
	// ERC-20 allowance value.
	maxUint256 = new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 256), big.NewInt(1))
	// maxUint160 is type(uint160).max — Permit2 amounts are uint160,
	// so this is its "unlimited" sentinel.
	maxUint160 = new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 160), big.NewInt(1))
)

// DomainChainID returns the EIP-712 domain's chainId normalised to a
// decimal string. ok is false when the domain omits chainId (it is an
// optional field) or the value can't be parsed.
func (td *EIP712TypedData) DomainChainID() (string, bool) {
	if td.Domain == nil {
		return "", false
	}
	switch v := td.Domain["chainId"].(type) {
	case string:
		if v == "" {
			return "", false
		}
		s, base := v, 10
		if strings.HasPrefix(s, "0x") || strings.HasPrefix(s, "0X") {
			s, base = s[2:], 16
		}
		if b, ok := new(big.Int).SetString(s, base); ok {
			return b.Text(10), true
		}
	case float64:
		return strconv.FormatInt(int64(v), 10), true
	case int:
		return strconv.Itoa(v), true
	case int64:
		return strconv.FormatInt(v, 10), true
	case json.Number:
		if b, ok := new(big.Int).SetString(string(v), 10); ok {
			return b.Text(10), true
		}
	}
	return "", false
}

// SecurityWarnings inspects parsed typed data for cross-chain replay
// and dangerous-approval risks (ERC-2612 Permit, Permit2
// PermitSingle/PermitBatch, token approve/increaseAllowance, Seaport
// orders) and returns additive advisories for the approval sheet.
// activeChainID is the wallet's current EVM network chain id (decimal
// string); pass "" to skip the cross-chain check. These warnings never
// block signing — they exist purely to defeat blind signing by
// surfacing spender / amount / deadline, explicitly flagging unlimited
// (type(uint256).max) allowances.
func (td *EIP712TypedData) SecurityWarnings(activeChainID string) []SignWarning {
	var out []SignWarning

	// (a) Cross-chain: a signature whose domain targets a different
	// chain than the wallet is on can be replayed on that chain.
	if activeChainID != "" {
		if dom, ok := td.DomainChainID(); ok && dom != activeChainID {
			out = append(out, SignWarning{
				Code:     "eip712_chain_mismatch",
				Severity: "warn",
				Message:  fmt.Sprintf("This typed-data signature targets chain %s but your wallet is on chain %s. It may be replayable on the other chain.", dom, activeChainID),
				Field:    "chainId",
			})
		}
	}

	// (b) Dangerous primitives.
	msg := td.Message
	ptl := strings.ToLower(td.PrimaryType)

	appendAllowance := func(code, spender string, amount *big.Int, deadline string) {
		unlimited := amount != nil && (amount.Cmp(maxUint256) == 0 || amount.Cmp(maxUint160) == 0)
		var b strings.Builder
		b.WriteString("Signing this authorises a token spending approval")
		if spender != "" {
			b.WriteString(" to spender " + spender)
		}
		if amount != nil {
			if unlimited {
				b.WriteString(" for an UNLIMITED amount")
			} else {
				b.WriteString(" for amount " + amount.String())
			}
		}
		if deadline != "" {
			b.WriteString(" (deadline " + deadline + ")")
		}
		b.WriteString(". Only sign if you trust this site.")
		if unlimited {
			code += "_unlimited"
		}
		out = append(out, SignWarning{Code: code, Severity: "warn", Message: b.String(), Field: "spender"})
	}

	switch {
	case ptl == "permit":
		// ERC-2612: { owner, spender, value, nonce, deadline }.
		appendAllowance("eip712_permit", eip712Str(msg, "spender"),
			eip712Big(msg, "value"), eip712Display(msg, "deadline"))
	case ptl == "permitsingle":
		// Permit2 PermitSingle: { details:{token,amount,...}, spender, sigDeadline }.
		var amount *big.Int
		if d, ok := msg["details"].(map[string]any); ok {
			amount = eip712Big(d, "amount")
		}
		appendAllowance("permit2_approve", eip712Str(msg, "spender"), amount,
			eip712Display(msg, "sigDeadline"))
	case ptl == "permitbatch":
		// Permit2 PermitBatch: { details:[{token,amount,...}], spender, sigDeadline }.
		spender := eip712Str(msg, "spender")
		deadline := eip712Display(msg, "sigDeadline")
		if arr, ok := msg["details"].([]any); ok && len(arr) > 0 {
			for _, it := range arr {
				if d, ok := it.(map[string]any); ok {
					appendAllowance("permit2_approve", spender, eip712Big(d, "amount"), deadline)
				}
			}
		} else {
			appendAllowance("permit2_approve", spender, nil, deadline)
		}
	case strings.Contains(ptl, "increaseallowance"):
		amount := eip712Big(msg, "value")
		if amount == nil {
			amount = eip712Big(msg, "amount")
		}
		appendAllowance("eip712_increase_allowance", eip712Str(msg, "spender"), amount, "")
	case strings.Contains(ptl, "approve"):
		amount := eip712Big(msg, "value")
		if amount == nil {
			amount = eip712Big(msg, "amount")
		}
		appendAllowance("eip712_token_approval", eip712Str(msg, "spender"), amount, "")
	case strings.Contains(ptl, "order"):
		// Seaport (OpenSea) OrderComponents and similar marketplace
		// orders — signing can list or transfer your NFTs/tokens.
		out = append(out, SignWarning{
			Code:     "seaport_order",
			Severity: "warn",
			Message:  "This signs a marketplace order (Seaport-style) that can list or transfer your NFTs/tokens. Verify the collection, price, and recipient before signing.",
		})
	default:
		// Message-level match: a custom primaryType wrapping an
		// approval still exposes a spender + amount the user should
		// see.
		if msg != nil {
			if _, ok := msg["spender"]; ok {
				amount := eip712Big(msg, "value")
				if amount == nil {
					amount = eip712Big(msg, "amount")
				}
				appendAllowance("eip712_token_approval", eip712Str(msg, "spender"), amount,
					eip712Display(msg, "deadline"))
			}
		}
	}

	return out
}

// eip712Str returns m[key] as a string, or "" when absent / not a string.
func eip712Str(m map[string]any, key string) string {
	if m == nil {
		return ""
	}
	if s, ok := m[key].(string); ok {
		return s
	}
	return ""
}

// eip712Big parses m[key] (string decimal/hex, JSON number, or float)
// into a big.Int. Returns nil when absent or unparsable.
func eip712Big(m map[string]any, key string) *big.Int {
	if m == nil {
		return nil
	}
	switch v := m[key].(type) {
	case string:
		s, base := v, 10
		if strings.HasPrefix(s, "0x") || strings.HasPrefix(s, "0X") {
			s, base = s[2:], 16
		}
		if b, ok := new(big.Int).SetString(s, base); ok {
			return b
		}
	case float64:
		return new(big.Int).SetInt64(int64(v))
	case json.Number:
		if b, ok := new(big.Int).SetString(string(v), 10); ok {
			return b
		}
	}
	return nil
}

// eip712Display renders m[key] for human-readable warning text.
func eip712Display(m map[string]any, key string) string {
	if m == nil {
		return ""
	}
	switch v := m[key].(type) {
	case string:
		return v
	case float64:
		return strconv.FormatInt(int64(v), 10)
	case json.Number:
		return string(v)
	}
	return ""
}

// EIP712TypedData represents the full EIP-712 typed data structure.
type EIP712TypedData struct {
	Types       map[string][]EIP712Field `json:"types"`
	PrimaryType string                   `json:"primaryType"`
	Domain      map[string]any           `json:"domain"`
	Message     map[string]any           `json:"message"`
}

// EIP712Field represents a single field in an EIP-712 type definition.
type EIP712Field struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

// HashEIP712 computes the EIP-712 digest: keccak256("\x19\x01" || domainSeparator || hashStruct(message))
func (td *EIP712TypedData) HashEIP712() ([]byte, error) {
	domainSep, err := td.hashStruct("EIP712Domain", td.Domain)
	if err != nil {
		return nil, fmt.Errorf("domain separator: %w", err)
	}

	msgHash, err := td.hashStruct(td.PrimaryType, td.Message)
	if err != nil {
		return nil, fmt.Errorf("message hash: %w", err)
	}

	// "\x19\x01" || domainSeparator || hashStruct(message)
	raw := []byte{0x19, 0x01}
	raw = append(raw, domainSep...)
	raw = append(raw, msgHash...)
	return keccak256(raw), nil
}

// hashStruct computes keccak256(typeHash || encodeData(value))
func (td *EIP712TypedData) hashStruct(typeName string, data map[string]any) ([]byte, error) {
	typeHash, err := td.typeHash(typeName)
	if err != nil {
		return nil, err
	}

	encoded, err := td.encodeData(typeName, data)
	if err != nil {
		return nil, err
	}

	raw := make([]byte, 0, 32+len(encoded))
	raw = append(raw, typeHash...)
	raw = append(raw, encoded...)
	return keccak256(raw), nil
}

// typeHash returns keccak256(encodeType(typeName))
func (td *EIP712TypedData) typeHash(typeName string) ([]byte, error) {
	enc, err := td.encodeType(typeName)
	if err != nil {
		return nil, err
	}
	return keccak256([]byte(enc)), nil
}

// encodeType returns the type encoding string per EIP-712.
// e.g. "Mail(address from,address to,string contents)"
// Referenced types are appended in sorted order.
func (td *EIP712TypedData) encodeType(typeName string) (string, error) {
	fields, ok := td.Types[typeName]
	if !ok {
		return "", fmt.Errorf("type %s not found", typeName)
	}

	// Collect referenced struct types
	deps := make(map[string]bool)
	td.findDeps(typeName, deps)
	delete(deps, typeName) // primary type comes first

	sortedDeps := make([]string, 0, len(deps))
	for d := range deps {
		sortedDeps = append(sortedDeps, d)
	}
	sort.Strings(sortedDeps)

	result := formatType(typeName, fields)
	for _, dep := range sortedDeps {
		result += formatType(dep, td.Types[dep])
	}
	return result, nil
}

func formatType(name string, fields []EIP712Field) string {
	parts := make([]string, len(fields))
	for i, f := range fields {
		parts[i] = f.Type + " " + f.Name
	}
	return name + "(" + strings.Join(parts, ",") + ")"
}

// findDeps recursively finds all struct type dependencies.
func (td *EIP712TypedData) findDeps(typeName string, deps map[string]bool) {
	if deps[typeName] {
		return
	}
	fields, ok := td.Types[typeName]
	if !ok {
		return
	}
	deps[typeName] = true
	for _, f := range fields {
		baseType := stripArraySuffix(f.Type)
		if _, isStruct := td.Types[baseType]; isStruct {
			td.findDeps(baseType, deps)
		}
	}
}

// encodeData encodes the values of a struct according to EIP-712.
func (td *EIP712TypedData) encodeData(typeName string, data map[string]any) ([]byte, error) {
	fields, ok := td.Types[typeName]
	if !ok {
		return nil, fmt.Errorf("type %s not found", typeName)
	}

	var encoded []byte
	for _, field := range fields {
		val := data[field.Name]
		enc, err := td.encodeValue(field.Type, val)
		if err != nil {
			return nil, fmt.Errorf("field %s.%s: %w", typeName, field.Name, err)
		}
		encoded = append(encoded, enc...)
	}
	return encoded, nil
}

// encodeValue encodes a single value per EIP-712 rules.
func (td *EIP712TypedData) encodeValue(typ string, val any) ([]byte, error) {
	// Array types — both dynamic (T[]) and fixed-size (T[N]). EIP-712
	// encodes either as keccak256 of the concatenated member encodings.
	if elemType, fixedLen, isArray := arrayElemType(typ); isArray {
		arr, ok := val.([]any)
		if !ok {
			return nil, fmt.Errorf("expected array for %s", typ)
		}
		if fixedLen >= 0 && len(arr) != fixedLen {
			return nil, fmt.Errorf("%s expects %d elements, got %d", typ, fixedLen, len(arr))
		}
		var inner []byte
		for _, item := range arr {
			enc, err := td.encodeValue(elemType, item)
			if err != nil {
				return nil, err
			}
			inner = append(inner, enc...)
		}
		return keccak256(inner), nil
	}

	// Struct types (referenced types)
	if _, isStruct := td.Types[typ]; isStruct {
		m, ok := val.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("expected object for struct type %s", typ)
		}
		h, err := td.hashStruct(typ, m)
		if err != nil {
			return nil, err
		}
		return h, nil
	}

	// Atomic types
	switch {
	case typ == "string":
		s, _ := val.(string)
		return keccak256([]byte(s)), nil
	case typ == "bytes":
		s, ok := val.(string)
		if !ok {
			return nil, errors.New("bytes value must be hex string")
		}
		b, err := hexDecode(s)
		if err != nil {
			return nil, fmt.Errorf("bytes: %w", err)
		}
		return keccak256(b), nil
	case typ == "bool":
		return padLeft32(boolToBytes(val)), nil
	case typ == "address":
		s, _ := val.(string)
		b, err := hexDecode(s)
		if err != nil {
			return nil, fmt.Errorf("address: %w", err)
		}
		return padLeft32(b), nil
	case strings.HasPrefix(typ, "uint"):
		n, ok := bigIntFromVal(val)
		if !ok {
			return nil, fmt.Errorf("invalid value for %s", typ)
		}
		if n.Sign() < 0 {
			return nil, fmt.Errorf("negative value for unsigned %s", typ)
		}
		return encodeUint256(n), nil
	case strings.HasPrefix(typ, "int"):
		n, ok := bigIntFromVal(val)
		if !ok {
			return nil, fmt.Errorf("invalid value for %s", typ)
		}
		return encodeInt256(n), nil
	case strings.HasPrefix(typ, "bytes"):
		// Fixed-size bytesN
		s, _ := val.(string)
		b, err := hexDecode(s)
		if err != nil {
			return nil, fmt.Errorf("%s: %w", typ, err)
		}
		return padRight32(b), nil
	default:
		return nil, fmt.Errorf("unsupported EIP-712 type: %s", typ)
	}
}

func stripArraySuffix(t string) string {
	if idx := strings.Index(t, "["); idx != -1 {
		return t[:idx]
	}
	return t
}

func keccak256(data []byte) []byte {
	h := sha3.NewLegacyKeccak256()
	h.Write(data)
	return h.Sum(nil)
}

func padLeft32(b []byte) []byte {
	if len(b) >= 32 {
		return b[len(b)-32:]
	}
	padded := make([]byte, 32)
	copy(padded[32-len(b):], b)
	return padded
}

func padRight32(b []byte) []byte {
	if len(b) >= 32 {
		return b[:32]
	}
	padded := make([]byte, 32)
	copy(padded, b)
	return padded
}

// arrayElemType reports whether typ is an array type and, if so,
// returns its element type and fixed length. fixedLen is -1 for a
// dynamic array (T[]) and >= 0 for a fixed-size array (T[N]).
func arrayElemType(typ string) (elem string, fixedLen int, ok bool) {
	if !strings.HasSuffix(typ, "]") {
		return "", 0, false
	}
	open := strings.LastIndex(typ, "[")
	if open < 0 {
		return "", 0, false
	}
	inner := typ[open+1 : len(typ)-1]
	base := typ[:open]
	if inner == "" {
		return base, -1, true
	}
	n, err := strconv.Atoi(inner)
	if err != nil || n < 0 {
		return "", 0, false
	}
	return base, n, true
}

// hexDecode decodes a 0x-prefixed (or bare) hex string, returning an
// error on invalid input instead of silently zero-filling — so a
// malformed value aborts signing rather than producing an attacker-
// chosen (zero) digest input.
func hexDecode(s string) ([]byte, error) {
	s = strings.TrimPrefix(s, "0x")
	s = strings.TrimPrefix(s, "0X")
	if s == "" {
		return []byte{}, nil
	}
	if len(s)%2 != 0 {
		s = "0" + s
	}
	return hex.DecodeString(s)
}

// bigIntFromVal parses an EIP-712 numeric value (decimal/hex string,
// JSON number, or float) into a big.Int, preserving sign.
func bigIntFromVal(val any) (*big.Int, bool) {
	switch v := val.(type) {
	case string:
		v = strings.TrimSpace(v)
		neg := strings.HasPrefix(v, "-")
		if neg {
			v = v[1:]
		}
		n := new(big.Int)
		var ok bool
		if strings.HasPrefix(v, "0x") || strings.HasPrefix(v, "0X") {
			_, ok = n.SetString(v[2:], 16)
		} else {
			_, ok = n.SetString(v, 10)
		}
		if !ok {
			return nil, false
		}
		if neg {
			n.Neg(n)
		}
		return n, true
	case json.Number:
		n := new(big.Int)
		if _, ok := n.SetString(string(v), 10); ok {
			return n, true
		}
		return nil, false
	case float64:
		return new(big.Int).SetInt64(int64(v)), true
	default:
		return nil, false
	}
}

// encodeUint256 renders the low 256 bits of n as big-endian 32 bytes.
func encodeUint256(n *big.Int) []byte {
	b := make([]byte, 32)
	new(big.Int).And(n, maxUint256).FillBytes(b)
	return b
}

// encodeInt256 renders n as a 32-byte two's-complement big-endian
// integer — correct for negatives and for positive magnitudes whose
// top bit is set (which the previous magnitude+sign-extend code got
// wrong).
func encodeInt256(n *big.Int) []byte {
	b := make([]byte, 32)
	m := new(big.Int)
	if n.Sign() >= 0 {
		m.And(n, maxUint256)
	} else {
		// 2^256 + n, masked to 256 bits.
		m.Add(new(big.Int).Lsh(big.NewInt(1), 256), n)
		m.And(m, maxUint256)
	}
	m.FillBytes(b)
	return b
}

func boolToBytes(val any) []byte {
	switch v := val.(type) {
	case bool:
		if v {
			return []byte{1}
		}
		return []byte{0}
	default:
		return []byte{0}
	}
}

// ParseEIP712TypedData parses a JSON string into EIP712TypedData.
func ParseEIP712TypedData(data string) (*EIP712TypedData, error) {
	var td EIP712TypedData
	if err := json.Unmarshal([]byte(data), &td); err != nil {
		return nil, fmt.Errorf("failed to parse typed data: %w", err)
	}
	if td.PrimaryType == "" {
		return nil, errors.New("primaryType is required")
	}
	if td.Types == nil {
		return nil, errors.New("types is required")
	}
	if _, ok := td.Types[td.PrimaryType]; !ok {
		return nil, fmt.Errorf("primaryType %s not found in types", td.PrimaryType)
	}
	return &td, nil
}

File: rust/src/eip712.rs
//! EIP-712 typed-structured-data hashing (port of `wltbase/eip712.go`). Produces
//! the signable digest `keccak256("\x19\x01" || domainSeparator ||
//! hashStruct(message))` that `eth_signTypedData_v3/v4` signs.

use std::collections::{BTreeSet, HashMap};

use num_bigint::{BigInt, Sign};
use serde::Deserialize;
use serde_json::{Map, Value};

use purecrypto::hash::keccak256;

#[derive(Deserialize, Debug, Clone)]
pub struct TypedData {
    #[serde(default)]
    pub types: HashMap<String, Vec<Field>>,
    #[serde(rename = "primaryType", default)]
    pub primary_type: String,
    #[serde(default)]
    pub domain: Map<String, Value>,
    #[serde(default)]
    pub message: Map<String, Value>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct Field {
    pub name: String,
    #[serde(rename = "type")]
    pub typ: String,
}

/// Parse a typed-data JSON string (Go `ParseEIP712TypedData`).
pub fn parse(data: &str) -> Result<TypedData, String> {
    let td: TypedData = serde_json::from_str(data).map_err(|e| format!("failed to parse typed data: {e}"))?;
    if td.primary_type.is_empty() {
        return Err("primaryType is required".into());
    }
    if td.types.is_empty() {
        return Err("types is required".into());
    }
    if !td.types.contains_key(&td.primary_type) {
        return Err(format!("primaryType {} not found in types", td.primary_type));
    }
    Ok(td)
}

impl TypedData {
    /// The EIP-712 signable digest: `keccak256("\x19\x01" || domainSep || hashStruct(message))`.
    pub fn hash(&self) -> Result<[u8; 32], String> {
        let domain_sep = self.hash_struct("EIP712Domain", &self.domain)?;
        let msg_hash = self.hash_struct(&self.primary_type, &self.message)?;
        let mut raw = vec![0x19u8, 0x01];
        raw.extend_from_slice(&domain_sep);
        raw.extend_from_slice(&msg_hash);
        Ok(keccak256(&raw))
    }

    /// `keccak256(typeHash || encodeData(value))`.
    fn hash_struct(&self, type_name: &str, data: &Map<String, Value>) -> Result<[u8; 32], String> {
        let type_hash = self.type_hash(type_name)?;
        let encoded = self.encode_data(type_name, data)?;
        let mut raw = Vec::with_capacity(32 + encoded.len());
        raw.extend_from_slice(&type_hash);
        raw.extend_from_slice(&encoded);
        Ok(keccak256(&raw))
    }

    fn type_hash(&self, type_name: &str) -> Result<[u8; 32], String> {
        Ok(keccak256(self.encode_type(type_name)?.as_bytes()))
    }

    /// The type encoding string, e.g. `Mail(address from,...)Person(...)` with
    /// referenced types appended in sorted order.
    fn encode_type(&self, type_name: &str) -> Result<String, String> {
        let fields = self.types.get(type_name).ok_or_else(|| format!("type {type_name} not found"))?;
        let mut deps = BTreeSet::new();
        self.find_deps(type_name, &mut deps);
        deps.remove(type_name); // primary type comes first
        let mut result = format_type(type_name, fields);
        for dep in &deps {
            result.push_str(&format_type(dep, &self.types[dep]));
        }
        Ok(result)
    }

    fn find_deps(&self, type_name: &str, deps: &mut BTreeSet<String>) {
        if deps.contains(type_name) {
            return;
        }
        let Some(fields) = self.types.get(type_name) else { return };
        deps.insert(type_name.to_owned());
        for f in fields {
            let base = strip_array_suffix(&f.typ);
            if self.types.contains_key(base) {
                self.find_deps(base, deps);
            }
        }
    }

    fn encode_data(&self, type_name: &str, data: &Map<String, Value>) -> Result<Vec<u8>, String> {
        let fields = self.types.get(type_name).ok_or_else(|| format!("type {type_name} not found"))?;
        let mut encoded = Vec::new();
        for field in fields {
            let val = data.get(&field.name).unwrap_or(&Value::Null);
            let enc = self
                .encode_value(&field.typ, val)
                .map_err(|e| format!("field {type_name}.{}: {e}", field.name))?;
            encoded.extend_from_slice(&enc);
        }
        Ok(encoded)
    }

    fn encode_value(&self, typ: &str, val: &Value) -> Result<Vec<u8>, String> {
        // Array types (T[] / T[N]) → keccak of concatenated member encodings.
        if let Some((elem, fixed_len)) = array_elem_type(typ) {
            let arr = val.as_array().ok_or_else(|| format!("expected array for {typ}"))?;
            if let Some(n) = fixed_len {
                if arr.len() != n {
                    return Err(format!("{typ} expects {n} elements, got {}", arr.len()));
                }
            }
            let mut inner = Vec::new();
            for item in arr {
                inner.extend_from_slice(&self.encode_value(&elem, item)?);
            }
            return Ok(keccak256(&inner).to_vec());
        }
        // Referenced struct types → hashStruct.
        if self.types.contains_key(typ) {
            let m = val.as_object().ok_or_else(|| format!("expected object for struct type {typ}"))?;
            return Ok(self.hash_struct(typ, m)?.to_vec());
        }
        // Atomic types.
        match typ {
            "string" => Ok(keccak256(val.as_str().unwrap_or("").as_bytes()).to_vec()),
            "bytes" => {
                let s = val.as_str().ok_or("bytes value must be hex string")?;
                Ok(keccak256(&hex_decode(s)?).to_vec())
            }
            "bool" => Ok(pad_left32(&[bool_to_byte(val)])),
            "address" => {
                let s = val.as_str().unwrap_or("");
                Ok(pad_left32(&hex_decode(s)?))
            }
            _ if typ.starts_with("uint") => {
                let n = bigint_from_val(val).ok_or_else(|| format!("invalid value for {typ}"))?;
                if n.sign() == Sign::Minus {
                    return Err(format!("negative value for unsigned {typ}"));
                }
                Ok(encode_uint256(&n))
            }
            _ if typ.starts_with("int") => {
                let n = bigint_from_val(val).ok_or_else(|| format!("invalid value for {typ}"))?;
                Ok(encode_int256(&n))
            }
            _ if typ.starts_with("bytes") => {
                // Fixed-size bytesN — right-padded.
                let s = val.as_str().unwrap_or("");
                Ok(pad_right32(&hex_decode(s)?))
            }
            other => Err(format!("unsupported EIP-712 type: {other}")),
        }
    }
}

fn format_type(name: &str, fields: &[Field]) -> String {
    let parts: Vec<String> = fields.iter().map(|f| format!("{} {}", f.typ, f.name)).collect();
    format!("{name}({})", parts.join(","))
}

fn strip_array_suffix(t: &str) -> &str {
    match t.find('[') {
        Some(i) => &t[..i],
        None => t,
    }
}

/// `(elemType, Some(n)|None)` for a fixed/dynamic array type, else None.
fn array_elem_type(typ: &str) -> Option<(String, Option<usize>)> {
    if !typ.ends_with(']') {
        return None;
    }
    let open = typ.rfind('[')?;
    let inner = &typ[open + 1..typ.len() - 1];
    let base = typ[..open].to_owned();
    if inner.is_empty() {
        return Some((base, None));
    }
    inner.parse::<usize>().ok().map(|n| (base, Some(n)))
}

/// Decode a 0x-prefixed (or bare) hex string, erroring on invalid input.
fn hex_decode(s: &str) -> Result<Vec<u8>, String> {
    let s = s.strip_prefix("0x").or_else(|| s.strip_prefix("0X")).unwrap_or(s);
    if s.is_empty() {
        return Ok(Vec::new());
    }
    let s = if s.len() % 2 != 0 { format!("0{s}") } else { s.to_owned() };
    (0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).map_err(|e| e.to_string()))
        .collect()
}

/// Parse an EIP-712 numeric value (decimal/hex string or JSON number).
fn bigint_from_val(val: &Value) -> Option<BigInt> {
    match val {
        Value::String(s) => {
            let s = s.trim();
            let (neg, s) = match s.strip_prefix('-') {
                Some(rest) => (true, rest),
                None => (false, s),
            };
            let n = if let Some(h) = s.strip_prefix("0x").or_else(|| s.strip_prefix("0X")) {
                BigInt::parse_bytes(h.as_bytes(), 16)?
            } else {
                BigInt::parse_bytes(s.as_bytes(), 10)?
            };
            Some(if neg { -n } else { n })
        }
        Value::Number(n) => BigInt::parse_bytes(n.to_string().as_bytes(), 10),
        _ => None,
    }
}

fn max_uint256() -> BigInt {
    (BigInt::from(1) << 256) - 1
}

/// The low 256 bits of `n` as big-endian 32 bytes.
fn encode_uint256(n: &BigInt) -> Vec<u8> {
    let masked = n & max_uint256();
    to_be_32(&masked)
}

/// `n` as a 32-byte two's-complement big-endian integer.
fn encode_int256(n: &BigInt) -> Vec<u8> {
    let m = if n.sign() != Sign::Minus {
        n & max_uint256()
    } else {
        ((BigInt::from(1) << 256) + n) & max_uint256()
    };
    to_be_32(&m)
}

/// A non-negative BigInt (< 2^256) as exactly 32 big-endian bytes.
fn to_be_32(n: &BigInt) -> Vec<u8> {
    let (_, mut bytes) = n.to_bytes_be();
    if bytes.len() > 32 {
        bytes = bytes[bytes.len() - 32..].to_vec();
    }
    let mut out = vec![0u8; 32 - bytes.len()];
    out.extend_from_slice(&bytes);
    out
}

fn pad_left32(b: &[u8]) -> Vec<u8> {
    if b.len() >= 32 {
        return b[b.len() - 32..].to_vec();
    }
    let mut out = vec![0u8; 32 - b.len()];
    out.extend_from_slice(b);
    out
}

fn pad_right32(b: &[u8]) -> Vec<u8> {
    let mut out = if b.len() >= 32 { b[..32].to_vec() } else { b.to_vec() };
    out.resize(32, 0);
    out
}

fn bool_to_byte(val: &Value) -> u8 {
    matches!(val, Value::Bool(true)) as u8
}

#[cfg(test)]
mod tests {
    use super::*;

    // The canonical EIP-712 spec "Mail" example → its well-known signable digest.
    const MAIL: &str = r#"{
        "types": {
            "EIP712Domain": [
                {"name":"name","type":"string"},
                {"name":"version","type":"string"},
                {"name":"chainId","type":"uint256"},
                {"name":"verifyingContract","type":"address"}
            ],
            "Person": [
                {"name":"name","type":"string"},
                {"name":"wallet","type":"address"}
            ],
            "Mail": [
                {"name":"from","type":"Person"},
                {"name":"to","type":"Person"},
                {"name":"contents","type":"string"}
            ]
        },
        "primaryType": "Mail",
        "domain": {
            "name": "Ether Mail",
            "version": "1",
            "chainId": 1,
            "verifyingContract": "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC"
        },
        "message": {
            "from": {"name":"Cow","wallet":"0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"},
            "to": {"name":"Bob","wallet":"0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB"},
            "contents": "Hello, Bob!"
        }
    }"#;

    fn hex(b: &[u8]) -> String {
        b.iter().map(|x| format!("{x:02x}")).collect()
    }

    #[test]
    fn mail_example_digest_matches_spec() {
        let td = parse(MAIL).unwrap();
        // encodeType(Mail) with the sorted Person dependency appended.
        assert_eq!(td.encode_type("Mail").unwrap(), "Mail(Person from,Person to,string contents)Person(string name,address wallet)");
        // The domain separator and message hash are the spec's known values.
        assert_eq!(hex(&td.hash_struct("EIP712Domain", &td.domain).unwrap()), "f2cee375fa42b42143804025fc449deafd50cc031ca257e0b194a650a912090f");
        // The full signable digest.
        assert_eq!(hex(&td.hash().unwrap()), "be609aee343fb3c4b28e1df9e632fca64fcfaede20f02e86244efddf30957bd2");
    }

    #[test]
    fn missing_primary_type_errors() {
        assert!(parse(r#"{"types":{"X":[]},"domain":{},"message":{}}"#).is_err());
    }
}
